### PR TITLE
Merge trainers

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR merges the specific `SageMakerTrainer` into the main `Trainer` to make all the scripts work directly with model parallelism. In passing, a few internal breaking changes:
- `is_sagemaker_distributed_available` is renamed to `is_sagemaker_dp_enabled` since it's about data parallelism and not specifically distributed training, and it's True when the user has activated it, not when it's merely "available"
- in the ParallelMode enum, the case `SAGEMAKER_DISTRIBUTED` is renamed as well (but it wasn't used anywhere).

Both only concern the internals of the library and no public API is breaking.

### CODE REVIEW

**Review by LysandreJik — APPROVED** [review]
Great! I like how the changes are structured in the Trainer. It's very simple to compare across different implementations (TPU, SageMaker), and makes readability (and the eventual debugging) very simple.

LGTM! Thanks for taking care of it.

File: src/transformers/training_args.py [hunk 1]
```diff
@@ -519,6 +525,10 @@ class TrainingArguments:
         default=False, metadata={"help": "Whether or not to skip adding of memory profiler reports to metrics."}
     )
     _n_gpu: int = field(init=False, repr=False, default=-1)
+    mp_parameters: str = field(
+        default="",
+        metadata={"help": "Used by the SageMaker launcher to send mp-specific args. Ignored in SageMakerTrainer"},
```
Comment (philschmid) [inline src/transformers/training_args.py:530]:
```suggestion
        metadata={"help": "Used by the SageMaker launcher to send mp-specific args. Ignored in Trainer"},
```

File: src/transformers/training_args.py [hunk 2]
```diff
@@ -730,8 +744,10 @@ def parallel_mode(self):
         """
         if is_torch_tpu_available():
             return ParallelMode.TPU
-        elif is_sagemaker_distributed_available():
-            return ParallelMode.SAGEMAKER_DISTRIBUTED
+        elif is_sagemaker_mp_enabled:
```
Comment (philschmid) [inline src/transformers/training_args.py:747]:
```suggestion
        elif is_sagemaker_mp_enabled():
```

**Comment by philschmid** [comment]
I tested run all our `sagemaker/tests` and a few additional `model_parallel` tests. ✅ 
I also tested everything with the upcoming `pytorch 1.7.1` image with the new `smp 1.3.0` (their model parallelism library)✅

What is still open for me is how does this behaves with `Seq2SeqTrainer`. Can users now use the `Seq2SeqTrainer` for model parallelism too? Data parallelism works already. Tested with `BART`

After we have merged the  `SageMakerTrainer` with the `Trainer` I would update the docs for sagemaker/model parallelism and the tests in tests/sagemaker.

**Review by philschmid — APPROVED** [review]

**Pull request merged**